r2k.c

```diff
@@ -426,11 +434,19 @@ static long io_ioctl (struct file *file, unsigned int cmd,
 			void *kaddr;
 			int bytes;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,9,2)
```

Comment:
Again, this is not correct. This should be defined in `r2k.h` as a macro.

If you take a look at http://elixir.free-electrons.com/ you can find multiple different definitions (note this are random versions not the exact version deprecating or having the change for the first time):

**v4.5** (8 args)
```C
long get_user_pages(struct task_struct *tsk, struct mm_struct *mm,
		    unsigned long start, unsigned long nr_pages,
		    int write, int force, struct page **pages,
		    struct vm_area_struct **vmas);
```

**v4.7.4** (8 args)
```C
long get_user_pages_remote(struct task_struct *tsk, struct mm_struct *mm,
			    unsigned long start, unsigned long nr_pages,
			    int write, int force, struct page **pages,
			    struct vm_area_struct **vmas);
```

**v4.9.59** (7 args)
```C

long get_user_pages_remote(struct task_struct *tsk, struct mm_struct *mm,
			    unsigned long start, unsigned long nr_pages,
			    unsigned int gup_flags, struct page **pages,
			    struct vm_area_struct **vmas);
```

**v4.10** (8 args)
```C
long get_user_pages_remote(struct task_struct *tsk, struct mm_struct *mm,
			    unsigned long start, unsigned long nr_pages,
			    unsigned int gup_flags, struct page **pages,
			    struct vm_area_struct **vmas, int *locked);
```

So there are at least 4 different variations. An example to fix this could be, remove `get_user_pages` macro at `r2k.h` and define something like this:

```C
#if LINUX_VERSION_CODE < KERNEL_VERSION(4,6,0)
#define get_user_pages_wrapper(tsk, mm, start, nr_pages,  write, force, pages, vmas, locked) \
                get_user_pages(tsk, mm, start, nr_pages,  write, force, pages, vmas) 
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,6,0) && LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0)
#define get_user_pages_wrapper(tsk, mm, start, nr_pages,  write, force, pages, vmas, locked) \
                get_user_pages_remote(tsk, mm, start, nr_pages,  write, force, pages, vmas)
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,9,0) && LINUX_VERSION_CODE < KERNEL_VERSION(4,10,0)
#define get_user_pages_wrapper(tsk, mm, start, nr_pages,  write, force, pages, vmas, locked) \
                get_user_pages_remote(tsk, mm, start, nr_pages,  write | force, pages, vmas)
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,10,0)
#define get_user_pages_wrapper(tsk, mm, start, nr_pages,  write, force, pages, vmas, locked) \
                get_user_pages_remote(tsk, mm, start, nr_pages,  write | force, pages, vmas, locked)
#endif
```

Please note that its just an example, not versions, neither arguments are correct. for example I didn't looked at what `gup_flags` parameter does, so the `write | force` for it is totally wrong.

Reply:
Fixed, hopefully the introduced comment is clear enough.